pkg/volume/portworx/portworx_util.go

```diff
@@ -0,0 +1,168 @@
+	}
+
+	specHandler := osdspec.NewSpecHandler()
+	spec, err := specHandler.SpecFromOpts(p.options.Parameters)
+	if err != nil {
+		return "", 0, nil, err
+	}
+	spec.Size = uint64(requestGB * 1024 * 1024 * 1024)
+	source := osdapi.Source{}
+	locator := osdapi.VolumeLocator{
+		Name:         p.options.PVName,
+		VolumeLabels: labels,
+	}
+	volumeID, err := client.Create(&locator, &source, spec)
+	if err != nil {
+		glog.V(2).Infof("Error creating Portworx Volume : %v", err)
+	}
+	return volumeID, requestGB, nil, err
+}
+
+// DeleteVolume deletes a Portworx volume
+func (util *PortworxVolumeUtil) DeleteVolume(d *portworxVolumeDeleter) error {
+	hostName := d.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Delete(d.volumeID)
+	if err != nil {
+		glog.V(2).Infof("Error deleting Portworx Volume (%v): %v", d.volName, err)
+		return err
+	}
+	return nil
+}
+
+// AttachVolume attaches a Portworx Volume
+func (util *PortworxVolumeUtil) AttachVolume(m *portworxVolumeMounter) (string, error) {
+	hostName := m.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return "", err
+	}
+
+	devicePath, err := client.Attach(m.volName)
+	if err != nil {
+		glog.V(2).Infof("Error attaching Portworx Volume (%v): %v", m.volName, err)
+		return "", err
+	}
+	return devicePath, nil
+}
+
+// DetachVolume detaches a Portworx Volume
+func (util *PortworxVolumeUtil) DetachVolume(u *portworxVolumeUnmounter) error {
+	hostName := u.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Detach(u.volName)
+	if err != nil {
+		glog.V(2).Infof("Error detaching Portworx Volume (%v): %v", u.volName, err)
+		return err
+	}
+	return nil
+}
+
+// MountVolume mounts a Portworx Volume on the specified mountPath
+func (util *PortworxVolumeUtil) MountVolume(m *portworxVolumeMounter, mountPath string) error {
+	hostName := m.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Mount(m.volName, mountPath)
+	if err != nil {
+		glog.V(2).Infof("Error mounting Portworx Volume (%v) on Path (%v): %v", m.volName, mountPath, err)
+		return err
+	}
+	return nil
+}
+
+// UnmountVolume unmounts a Portworx Volume
+func (util *PortworxVolumeUtil) UnmountVolume(u *portworxVolumeUnmounter, mountPath string) error {
+	hostName := u.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Unmount(u.volName, mountPath)
+	if err != nil {
+		glog.V(2).Infof("Error unmounting Portworx Volume (%v) on Path (%v): %v", u.volName, mountPath, err)
+		return err
+	}
+	return nil
+}
+
+func (util *PortworxVolumeUtil) osdClient(hostName string) (osdvolume.VolumeDriver, error) {
+	if util.portworxClient == nil {
+		driverClient, err := volumeclient.NewDriverClient("", pxdDriverName, osdDriverVersion)
```

Comment:
The client can talk on a port or a unix socket. The first argument is for the hostname. We are initialzing the client to talk only over a unix socket, hence it is empty. I will add a constant for no hostname instead of empty string.